Modules/_pickle.c

```diff
@@ -360,6 +360,81 @@ _Pickle_FastCall(PyObject *func, PyObject *obj)
 
 /*************************************************************************/
 
+static int
+init_method_ref(PyObject *self, _Py_Identifier *name,
+                PyObject **method, int *unbound)
+{
+    PyObject *func, *func2;
+
+    /* *method and *unbound should be consistent.  All refcount decrements
+       should be occurred after setting *unbound and *method. */
+    *unbound = 0;
+    Py_CLEAR(*method);
+
+    func = _PyObject_GetAttrId(self, name);
+    if (func == NULL) {
+        *unbound = 0;
+        Py_CLEAR(*method);
+        if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
+            return -1;
+        }
+        PyErr_Clear();
+        return 0;
+    }
+
+    /* Avoid creating a reference cycle (pickler -> bound method of pickler ->
+       -> pickler). */
+    if (PyMethod_Check(func) && PyMethod_GET_SELF(func) == self) {
+        /* bound Python method */
+        func2 = PyMethod_GET_FUNCTION(func);
+        Py_INCREF(func2);
+    }
+    else if (PyCFunction_Check(func) && PyCFunction_GET_SELF(func) == self) {
+        /* bound built-in method */
+        func2 = PyDescr_NewMethod(Py_TYPE(self),
+                                 ((PyCFunctionObject *)func)->m_ml);
+        if (func2 == NULL) {
+            Py_DECREF(func);
+            return -1;
+        }
+    }
+    else {
+        *unbound = 0;
+        Py_XSETREF(*method, func);
+        return 0;
+    }
+    *unbound = 1;
+    Py_XSETREF(*method, func2);
+    Py_DECREF(func);
+    return 0;
+}
+
+static PyObject *
+bound_method(PyObject *self, PyObject *func, int unbound)
+{
+    if (unbound) {
+        descrgetfunc f = func->ob_type->tp_descr_get;
+        if (f != NULL) {
+            return f(func, self, (PyObject *)Py_TYPE(self));
+        }
+    }
```

Comment:
Why does this fall through? May we return a non-bound method when `unbound` is true?

Reply:
Hmm, interesting question.